src/stan_math_backend/Stan_math_code_gen.ml

```diff
@@ -23,7 +23,72 @@ open Fmt
 open Expression_gen
 open Statement_gen
 
-type found_functor = {template: string; signature: string; defn: string}
+(* TODO: move to seperate file and code gen more like this with a structured type *)
+type template =
+  | Typename of string
+  | Require of string list * string
+  | Bool of string
+
+let pp_template ppf template =
+  match template with
+  | Typename t -> pf ppf "typename %s" t
+  | Bool s -> pf ppf "bool %s" s
+  | Require ((require_list : string list), (name : string)) ->
+      let full_require_length = List.length require_list - 1 in
+      if full_require_length = 0 then
+        pf ppf "%s" ((List.nth_exn require_list) 0 ^ "<" ^ name ^ ">")
+      else
+        let rec repper (s : string) (n : int) : string =
+          match n with 1 -> s | _ -> s ^ repper s (n - 1) in
+        (* Why can't I just use | full_require_length here?*)
+        let mappy (depth : int) (acc : string) (require : string) =
+          let base_req = acc ^ require ^ "<" ^ name ^ ">" in
+          match depth with
+          | 0 -> base_req ^ ", "
+          | _ ->
+              let next_bool_expr =
+                acc ^ require ^ "<"
+                ^ repper "stan::value_type_t<" depth
+                ^ name ^ repper ">" depth ^ ">" in
+              if full_require_length = depth then next_bool_expr
+              else next_bool_expr ^ ", " in
+        let multi_bool_expr = List.foldi ~f:mappy require_list ~init:"" in
+        pf ppf "%s" multi_bool_expr
+
+let pp_template_defaults ppf template =
+  match template with
+  | Require _ -> pf ppf "%a = nullptr" pp_template template
+  | _ -> pp_template ppf template
+
+let pp_template_requires ppf template_lst =
+  pf ppf "stan::require_all_t<%a>*" (list ~sep:comma pp_template) template_lst
+
+let pp_template_requires_defaults ppf (template : template list) =
+  pf ppf "%a = nullptr" pp_template_requires template
+
+let pp_templates ~defaults ppf templates =
+  let find_requires default_val template =
+    match template with Require _ -> default_val | _ -> not default_val in
+  let templates' = List.filter ~f:(find_requires false) templates in
+  let require_templates =
+    List.filter ~f:(fun x -> find_requires true x) templates in
```

Comment:
Obviously too early for a proper review but I just wanted to point out the existence of the super useful [`List.partition_tf`](https://ocaml.janestreet.com/ocaml-core/v0.14/doc/core_kernel/Core_kernel/List/index.html#val-partition_tf) function which does exactly this kind of thing

So you could write something like:

```ocaml
let templates', require_templates = List.partition_tf templates 
       ~f:(function Require -> false | _ -> true)
```

Reply:
Thanks! Lol yeah I knew there was some way to easily do this and forgot the name of it